### ISSUE TITLE
Transform input validation errors

### ISSUE THREAD

**Issue by royibernthal** [TypedProject/tsed#696]
Version: 5.32.1

I'd like to transform input validation (or converter deserialization) errors into my own format.

Is there some way for me to register a middleware between the error and returning it in the response?

Model:

```
export default class Credentials {
    
    @Required()
    email: string;

    @Required()
    password: string;

}
```

Controller:

```
@Controller('/auth')
export class AuthController {

    @Get('/signup')
    async signUp(@QueryParams() credentials: Credentials) {
        return 'test';
    }

}
```

If I'm not passing an email in the query params, the endpoint's response is a plain string as follows:
"Property email on class Credentials is required."

I'd like to be able to transform it into something like this:
{ "error": { "code": "field_missing", info: "email field is missing" } }

Should I add my own custom middleware and use ConvertersService/deserialize() manually, or is there a more out of the box solution?

**Comment by Romakita** [comment]
Hello @royibernthal 
You've an example on ajv: 
http://tsed.io/tutorials/ajv.html#installation
```typescript
import {ServerLoader, ServerSettings} from "@tsed/common";
import "@tsed/ajv"; // import ajv ts.ed module

@ServerSettings({
    rootDir: __dirname,
     ajv: {
       errorFormat: (error) => `At ${error.modelName}${error.dataPath}, value '${error.data}' ${error.message}`,
       options: {verbose: true}
    },
})
export class Server extends ServerLoader {

}
```

**Comment by Romakita** [comment]
You can also override the middleware responsible to send error: http://tsed.io/docs/middlewares/override/global-error-handler.html

**Comment by royibernthal** [comment]
Hi @Romakita, thanks for the quick response.

avj seems to only be responsible for part of the error displayed, I'm still getting a BadRequest string outputted above that error format. Correct me if I'm missing something.

I tried overriding the global error handler. It seems though as if my override is not registered.
I tried both logging and returning null. Both had no effect whatsoever.
I placed MyGEHMiddleware.ts in /middlewares.

```
import { Err, GlobalErrorHandlerMiddleware, OverrideProvider, Req, Res } from '@tsed/common';

@OverrideProvider(GlobalErrorHandlerMiddleware)
export class MyGEHMiddleware extends GlobalErrorHandlerMiddleware {
    
    use(@Err() error: any, @Req() request: Req, @Res() response: Res): any {
        console.log('---------------------------------------------------------call');

        const value = super.use(error, request, response);

        console.log('---------------------------------------------------------value', value);

        return null;
    }

}
```

**Comment by Romakita** [comment]
You have to import on GlobalErrorHandlerMiddleware in your server:

```typescript
import "./MyGEHMiddleware"

@ServerSettings()
export class Server extends ServerLoader {

}
```
It should works :)

**Comment by royibernthal** [comment]
Yes when I'm importing it manually it does use it :)
Shouldn't it be imported automatically if it's under /middlewares?

Regardless, now even when I'm literally copy pasting the sample override from the docs I'm getting an error.

http://tsed.io/docs/middlewares/override/global-error-handler.html

```
import {Err, GlobalErrorHandlerMiddleware, OverrideProvider, Req, Res} from "@tsed/common";

@OverrideProvider(GlobalErrorHandlerMiddleware)
export class MyGEHMiddleware extends GlobalErrorHandlerMiddleware {

  use(@Err() error: any,
      @Req() request: Req,
      @Res() response: Res): any {

    // DO SOMETHING

    return super.use(error, request, response);
  }
}
```

Error:

> TypeError: Cannot read property 'toLowerCase' of undefined
>     at ServerResponse.header (D:\Projects\testApp\server\node_modules\express\lib\response.js:761:15)
>     at MyGEHMiddleware.setHeaders (D:\Projects\testApp\server\node_modules\@tsed\common\src\server\components\GlobalErrorHandlerMiddleware.ts:72:16)
>     at MyGEHMiddleware.use (D:\Projects\testApp\server\node_modules\@tsed\common\src\server\components\GlobalErrorHandlerMiddleware.ts:27:12)
>     at MyGEHMiddleware.use (D:\Projects\testApp\server\build\middlewares\MyGEHMiddleware.js:19:22)
>     at HandlerBuilder.invoke (D:\Projects\testApp\server\node_modules\@tsed\common\src\mvc\builders\HandlerBuilder.ts:208:23)
>     at SafeSubscriber.rxjs_1.zip.pipe.subscribe.context [as _next] (D:\Projects\testApp\server\node_modules\@tsed\common\src\mvc\builders\HandlerBuilder.ts:182:34)
>     at SafeSubscriber.__tryOrUnsub (D:\Projects\testApp\server\node_modules\rxjs\src\internal\Subscriber.ts:265:10)
>     at SafeSubscriber.next (D:\Projects\testApp\server\node_modules\rxjs\src\internal\Subscriber.ts:207:14)
>     at Subscriber._next (D:\Projects\testApp\server\node_modules\rxjs\src\internal\Subscriber.ts:139:22)
>     at Subscriber.next (D:\Projects\testApp\server\node_modules\rxjs\src\internal\Subscriber.ts:99:12)
>     at MapSubscriber._next (D:\Projects\testApp\server\node_modules\rxjs\src\internal\operators\map.ts:89:22)
>     at MapSubscriber.Subscriber.next (D:\Projects\testApp\server\node_modules\rxjs\src\internal\Subscriber.ts:99:12)
>     at ZipSubscriber.checkIterators (D:\Projects\testApp\server\node_modules\rxjs\src\internal\observable\zip.ts:197:19)
>     at ZipBufferIterator.notifyNext (D:\Projects\testApp\server\node_modules\rxjs\src\internal\observable\zip.ts:324:17)
>     at InnerSubscriber._next (D:\Projects\testApp\server\node_modules\rxjs\src\internal\InnerSubscriber.ts:17:17)
>     at InnerSubscriber.Subscriber.next (D:\Projects\testApp\server\node_modules\rxjs\src\internal\Subscriber.ts:99:12)
>     at SwitchMapSubscriber.notifyNext (D:\Projects\testApp\server\node_modules\rxjs\src\internal\operators\switchMap.ts:166:24)
>     at InnerSubscriber._next (D:\Projects\testApp\server\node_modules\rxjs\src\internal\InnerSubscriber.ts:17:17)
>     at InnerSubscriber.Subscriber.next (D:\Projects\testApp\server\node_modules\rxjs\src\internal\Subscriber.ts:99:12)
>     at CatchSubscriber.Subscriber._next (D:\Projects\testApp\server\node_modules\rxjs\src\internal\Subscriber.ts:139:22)
>     at CatchSubscriber.Subscriber.next (D:\Projects\testApp\server\node_modules\rxjs\src\internal\Subscriber.ts:99:12)
>     at MapSubscriber._next (D:\Projects\testApp\server\node_modules\rxjs\src\internal\operators\map.ts:89:22)

**Comment by royibernthal** [comment]
@Romakita What do you think? :)

**Comment by royibernthal** [comment]
@Romakita I revisited this with fresh eyes, it's pretty obvious now that I look at it, I managed to get it working.

When there's a param check error (invalid params are passed to an endpoint), overriding this error handler and passing the params to super results in an error in setHeaders (due to error.origin being undefined?).

Maybe the params in the ovveriding version are not exactly the same params as the ones super receives without the override? (due to how the decorators work internally?)

Either way, as long as I intercept the error and don't pass it along to super nor manually call setHeaders manually, everything works.
